contracts/protocol/v1/proxies/P1SoloBridgeProxy.sol

```diff
@@ -0,0 +1,454 @@
+     */
+    function cancelTransfer(
+        Transfer calldata transfer
+    )
+        external
+    {
+        // Check permissions. Short-circuit if sender is the account owner.
+        if (msg.sender != transfer.account) {
+            I_Solo solo = I_Solo(_SOLO_MARGIN_);
+            I_PerpetualV1 perpetual = I_PerpetualV1(transfer.perpetual);
+            require(
+                _hasWithdrawPermissions(solo, perpetual, transfer),
+                "Sender does not have permission to cancel"
+            );
+        }
+
+        // Cancel the transfer.
+        bytes32 transferHash = _getTransferHash(transfer);
+        _HASH_USED_[transferHash] = true;
+
+        // Log the cancelation.
+        emit LogTransferCanceled(
+            transfer.account,
+            transferHash
+        );
+    }
+
+    // ============ Helper Functions ============
+
+    /**
+     * @dev Execute a withdrawal or deposit operation on Solo.
+     */
+    function _doSoloOperation(
+        I_Solo solo,
+        Transfer memory transfer,
+        bool isWithdrawal
+    )
+        private
+    {
+        // Create Solo account struct.
+        I_Solo.AccountInfo memory soloAccount = I_Solo.AccountInfo({
+            owner: transfer.account,
+            number: transfer.soloAccountNumber
+        });
+
+        // Create Solo accounts array.
+        I_Solo.AccountInfo[] memory soloAccounts = new I_Solo.AccountInfo[](1);
+        soloAccounts[0] = soloAccount;
+
+        // Create Solo actions array.
+        I_Solo.AssetAmount memory amount = I_Solo.AssetAmount({
+            sign: true,
+            denomination: I_Solo.AssetDenomination.Wei,
+            ref: I_Solo.AssetReference.Delta,
+            value: transfer.amount
+        });
+        I_Solo.ActionType actionType;
+        bytes memory data;
+        if (isWithdrawal) {
+            actionType = I_Solo.ActionType.Withdraw;
+            data = abi.encode(
+                I_Solo.WithdrawArgs({
+                    amount: amount,
+                    account: soloAccount,
+                    market: transfer.soloMarketId,
+                    to: address(this)
+                })
+            );
+        } else {
+            actionType = I_Solo.ActionType.Deposit;
+            data = abi.encode(
+                I_Solo.DepositArgs({
+                    amount: amount,
+                    account: soloAccount,
+                    market: transfer.soloMarketId,
+                    from: address(this)
+                })
+            );
+        }
+        I_Solo.ActionArgs[] memory soloActions = new I_Solo.ActionArgs[](1);
+        soloActions[0] = I_Solo.ActionArgs({
+            actionType: actionType,
+            accountId: transfer.soloAccountNumber,
+            amount: amount,
+            primaryMarketId: transfer.soloMarketId,
+            secondaryMarketId: 0,
+            otherAddress: address(0),
+            otherAccountId: 0,
+            data: data
```

Comment:
I think Solo actually ignores this data field for actions other than `buy/sell/trade/call`, so you can remove that logic all-together